### ISSUE TITLE
Allow (auto)switching front page between light mode and dark mode

### ISSUE THREAD

**Issue by sftim** [kubernetes/website#37444]
**This is a Feature Request**

**What would you like to be added**
A cut down form of https://github.com/kubernetes/website/issues/25061
Allow (auto)switching front page between light mode and dark mode

**Why is this needed**
Change https://kubernetes.io/ (**and only that page**) to support detecting the viewer's preference for either light mode or dark mode, and then honoring it.

This should use a similar technique to https://opensource.com/article/21/10/dark-themes-websites and must also play well with our custom Docsy theme.

For images, we might need a custom shortcode that renders to something like:
```html
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="/images/blocks-dark.svg">
  <source media="(prefers-color-scheme: light)" srcset="/images/blocks-light.svg">
 <img src="/images/blocks.svg" alt="">
</picture>
```

Implementing that shortcode or layout would be part of the work for this issue.

**Comments**
This is a minimum-viable-product approach. It's much easier to make the change work for just one page, and then iterate.

Future work might also allow a mechanism to let the user specify a preference for dark or light mode that is different from what their user agent (web browser) is asserting.

After covering https://kubernetes.io/ we could look at whether there are other pages on the site where we're happy to widen the automatic switching to. For example, https://kubernetes.io/community/

Alternatively, we might work on https://kubernetes.dev/ (full site) and then come back to https://kubernetes.io/ once we've learned how to get a site working with 2 color schemes. 

/area web-development
/kind feature

**Comment by sachinparihar** [comment]
/assign
I'd like to work on this.

**Comment by Ritikaa96** [comment]
Hi @sftim is anyone working on this feature? if not then I'd like to try adding this feature.

**Comment by sachinparihar** [comment]
> Hi @sftim is anyone working on this feature? if not then I'd like to try adding this feature.



> Hi @sftim is anyone working on this feature? if not then I'd like to try adding this feature.

I'm not working on this right now! if you want to work on this you can.

**Comment by tamilselvan1102** [comment]
@sftim 
suggest me where to place the toggle button.

![k8sHomePage](https://user-images.githubusercontent.com/20103020/202186094-bf4d0112-cae0-4da7-8b2a-de5f5ab3f300.JPG)

**Comment by sftim** [comment]
How about putting the button above "Edit this page"?
If you prefer another place, that's fine too.

**Comment by Ritikaa96** [comment]
For now i'd like to /assign it to myself and wait for a go ahead to work on it.

**Comment by sftim** [comment]
> For now i'd like to /assign it to myself and wait for a go ahead to work on it.

You (and all other contributors) don't need anyone's permission to start work. 
Anyone's welcome to work on this issue, or indeed any [k/website](https://github.com/kubernetes/website) issue. If someone has already made a start, it's polite to try to sync up and collaborate with them.

**Comment by Ritikaa96** [comment]
Thanks @sftim , 

> so _for now_ let's skip adding an override button, and rely on detecting the user's preference.

i am slightly confused by the statement. 
So as much I understood we are to add this feature to the one page _for now_ and no override button. 
Am i getting this right?
where do you suggest we put the changing action for view modes?

**Comment by sftim** [comment]
This is a limited scope issue to get familiar with using CSS to restyle the page, and to work out what we'll do with the images / icons.

> where do you suggest we put the changing action for view modes?

Someone who wants to try this out will have to set their browser to dark mode or light mode explicitly (or wait until dawn / sunset).

**Comment by Ritikaa96** [comment]
> Someone who wants to try this out will have to set their browser to dark mode or light mode explicitly (or wait until dawn / sunset).

Thanks @sftim , My first thought was that we want to add a button to help users choose which mode they prefer on the page.


> This is a limited scope issue to get familiar with using CSS to restyle the page, and to work out what we'll do with the images / icons.

I might have seen a reference to go through with this. will let you know if it work out right..

**Comment by Ritikaa96** [comment]
Hi @sftim I have written a shortcode for the <picture.. > it will look like this 

`{{< picture "light.svg" "dark.svg" "alt_text" >}}`

where we mention the images to be used in the preferred mode. 

are we looking for a similar solution?

**Comment by sftim** [comment]
> I have written a shortcode for the <picture.. > it will look like this
> 
> `{{< picture "light.svg" "dark.svg" "alt_text" >}}`
> 
> where we mention the images to be used in the preferred mode.

I might call it `{{< picture-light-and-dark "light.svg" "dark.svg" "alt_text" >}}` so that people remember which order.
We could also make our own custom extension to https://gohugo.io/content-management/shortcodes/#figure

**Comment by sftim** [comment]
It'd be great to see a preview of the effect, even if it's only a draft.

**Comment by Ritikaa96** [comment]
Sure let me comfirm first , we are to add this change in : https://github.com/kubernetes/website/blob/main/content/en/_index.html
Right?

**Comment by sftim** [comment]
https://github.com/kubernetes/website/blob/main/content/en/_index.html is one of the places we might need to change

I think the CSS (and its SCSS source code) is more important though.

**Comment by Ritikaa96** [comment]
> https://github.com/kubernetes/website/blob/main/content/en/_index.html is one of the places we might need to change
> 
> I think the CSS (and its SCSS source code) is more important.

Right. I have made some shortcode-related , also CSS related changes. but I might need some confirmation on these changes.  It's better if i confirm them here first.

**Comment by sftim** [comment]
Please [open a pull request](https://kubernetes.io/docs/contribute/new-content/open-a-pr/) to show the changes you're proposing. That's definitely the best way to collaborate here.

**Comment by Ritikaa96** [comment]
Hi @sftim , I'll make a PR and update here then.

**Comment by coder12git** [comment]
/assign

I am taking this, as i know how to implement this. I'll create a PR soon 🫡

**Comment by sftim** [comment]
@coder12git how's this work going?

**Comment by coder12git** [comment]
>Change https://kubernetes.io/ (and only that page) to support detecting the viewer's preference for either light mode or dark mode, and then honoring it.

This should use a similar technique to https://opensource.com/article/21/10/dark-themes-websites and must also play well with our custom Docsy theme.

@sftim As mentioned here just to do this for only one page, but if i follow the other point, then when i am doing changes it's happening to all other pages, i.e not able to implement this only for one page.

**Comment by tamilselvan1102** [comment]
@coder12git 
The layout is same for all the pages.  so if you add dark mode toggle button in header (Documentation,Kubernetes Blog,Training, ...) it will reflect all the pages. But if we add small logic in javascript, it is possible.

**Comment by tamilselvan1102** [comment]
let me do the try
/assign

**Comment by sftim** [comment]
>  The layout is same for all the pages. so if you add dark mode toggle button in header (Documentation,Kubernetes Blog,Training, ...) it will reflect all the pages. But if we add small logic in javascript, it is possible.

I recommend using CSS. [CSS is awesome](https://css-tricks.com/wp-content/uploads/2021/04/css-is-awesome.jpg) :wink:

See https://css-tricks.com/a-complete-guide-to-dark-mode-on-the-web/ for details, including approaches that don't require any JavaScript.

**Comment by tamilselvan1102** [comment]
i will use awesome icons

**Comment by sftim** [comment]
We discussed #45535 in today's SIG meeting.

**Comment by dipesh-rawat** [comment]
> In the long term, we might want to consider aligning with [Docsy](https://www.docsy.dev/) since it now supports [Dark Mode (announcement here)](https://www.docsy.dev/blog/2024/0.10.0/#color-themes-and-dark-mode-support).

xref https://github.com/kubernetes/website/issues/25061#issuecomment-2272352631

_Just cross-posting here for reference. The current progress on this issue is positive and looks great._

**Comment by Ant0wan** [comment]
Hi fellows,
Just for the curiosity - is there any ETA for this theme that will save my eyes ? :-) 
Thanks for having done it ! See ya around at some kube conf'

**Comment by sftim** [comment]
> ETA

No ETA I'm afraid. Most of the work on this website is done by volunteers.

**Comment by franciscojavierarceo** [comment]
Did @coder12git take this on?

**Comment by sftim** [comment]
> Did @coder12git take this on?

The [PR](https://github.com/kubernetes/website/pull/45535) we have (which is ready for review) is by @tamilselvan1102

**Comment by sftim** [comment]
See https://github.com/kubernetes/website/pull/49586 for the latest PR (reviewable)
